glean-core/python/glean/_dispatcher.py

```diff
@@ -2,11 +2,113 @@
+#     library used to interface with Rust releases the GIL around every foreign
+#     call. See https://cffi.readthedocs.io/en/latest/ref.html#conversions
+#
+#   - The other significant blocking I/O is in networking code, where the GIL
+#     is also released while waiting on network sockets.
+#
+# This approach greatly reduces complexity of the implementation. Using
+# multiprocessing would imply going to a 100% IPC-like approach, since the
+# Rust-side Glean objects could not be easily shared or message-passed across
+# the process boundary, whereas sharing across threads works transparently.
+
+
+log = logging.getLogger(__name__)
+
+
+class _ThreadWorker:
+    """
+    Manages a single worker to perform tasks in another thread.
+    """
+
+    def __init__(self):
+        self._queue = queue.Queue()
+        # The worker thread is only started when work needs to be performed so
+        # that importing Glean alone does not start an unnecessary thread.
+        self._started = False
+
+    def _add_task(self, sync: bool, task: Callable, *args, **kwargs):
+        """
+        Add a task to the worker queue.
+
+        Args:
+            sync (bool): If `True`, block until the task is complete.
+            task (Callable): The task to run.
+
+        Additional arguments are passed to the task.
+        """
+        if not self._started:
+            self._start_worker()
+        # If we are already on the worker thread, don't place the tasks in the
+        # queue, just run them now. This is required for synchronous testing
+        # mode, and also to run the tasks in the expected order.
+        if threading.get_ident() == self._ident:
+            task(*args, **kwargs)
+        else:
+            args = args or ()
+            kwargs = kwargs or {}
+            self._queue.put((task, args, kwargs))
+            if sync:
+                self._queue.join()
+
+    def add_task(self, sync: bool, task: Callable, *args, **kwargs):
+        """
+        Add a task to the worker queue.
+
+        Args:
+            sync (bool): If `True`, block until the task is complete.
+            task (Callable): The task to run.
+
+        Additional arguments are passed to the task.
+        """
+        self._add_task(sync, task, *args, **kwargs)
+
+    def _start_worker(self):
+        """
+        Starts the worker thread.
+
+        It is started in daemon mode so it will run as long as the parent
+        thread.
+        """
+        t = threading.Thread(target=self._worker)
+        t.daemon = True
+        t.start()
+        self._started = True
+        self._ident = t.ident
+
+    def _worker(self):
+        """
+        Implements the worker thread. Takes tasks off of the queue and runs
+        them.
+        """
+        while True:
+            task, args, kwargs = self._queue.get()
+            try:
+                task(*args, **kwargs)
+            except Exception as e:
+                log.error(str(e))
```

Comment:
```suggestion
                log.exception("Glean error")
```

See: https://www.loggly.com/blog/exceptional-logging-of-exceptions-in-python/

Reply:
TIL.  That's really cool.